File: backend-java/src/main/java/com/wealthtech/crm/modules/portfolioreview/service/EligibleFundExcelService.java
package com.wealthtech.crm.modules.portfolioreview.service;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.stereotype.Service;

import com.wealthtech.crm.modules.portfolioreview.dto.MasterFundRowDto;
import com.wealthtech.crm.modules.riskappetite.enums.ScoreCategory;
import com.wealthtech.crm.modules.usermanager.exception.BadRequestException;

import lombok.extern.slf4j.Slf4j;

/**
 * Service for generating master funds spreadsheet template and parsing uploaded Excel files.
 */
@Slf4j
@Service
public class EligibleFundExcelService {

    private static final String[] HEADERS = {
        "Fund Name", "ISIN", "Score Category", "Sub Category", "Asset Class", "Instrument Type", "Active (TRUE/FALSE)"
    };

    private final DataFormatter dataFormatter = new DataFormatter();

    /**
     * Generates a pre-styled Excel template with headers and representative sample records across risk appetite bands.
     */
    public byte[] generateMasterFundsTemplate() {
        try (Workbook workbook = new XSSFWorkbook();
             ByteArrayOutputStream out = new ByteArrayOutputStream()) {

            Sheet sheet = workbook.createSheet("Master Funds Template");

            // Header Style
            CellStyle headerStyle = workbook.createCellStyle();
            Font headerFont = workbook.createFont();
            headerFont.setBold(true);
            headerFont.setColor(IndexedColors.WHITE.getIndex());
            headerStyle.setFont(headerFont);
            headerStyle.setFillForegroundColor(IndexedColors.DARK_TEAL.getIndex());
            headerStyle.setFillPattern(FillPatternType.SOLID_FOREGROUND);
            headerStyle.setAlignment(HorizontalAlignment.CENTER);

            // Create Header Row
            Row headerRow = sheet.createRow(0);
            for (int i = 0; i < HEADERS.length; i++) {
                Cell cell = headerRow.createCell(i);
                cell.setCellValue(HEADERS[i]);
                cell.setCellStyle(headerStyle);
            }

            // Sample Data Rows across 5 Score Categories
            Object[][] sampleData = {
                {"HDFC Top 100 Fund - Direct Growth", "INF179K01BE2", "MODERATE", "Large Cap", "Equity", "Mutual Fund", "TRUE"},
                {"ICICI Prudential Liquid Fund - Direct Growth", "INF109K01BE1", "VERY_CONSERVATIVE", "Liquid", "Debt", "Mutual Fund", "TRUE"},
                {"Nippon India Small Cap Fund - Direct Growth", "INF204K01BE3", "VERY_AGGRESSIVE", "Small Cap", "Equity", "Mutual Fund", "TRUE"},
                {"Parag Parikh Flexi Cap Fund - Direct Growth", "INF879O01019", "AGGRESSIVE", "Flexi Cap", "Equity", "Mutual Fund", "TRUE"},
                {"SBI Balanced Advantage Fund - Direct Growth", "INF200K01BE4", "CONSERVATIVE", "Dynamic Asset Allocation", "Hybrid", "Mutual Fund", "TRUE"}
            };

            for (int r = 0; r < sampleData.length; r++) {
                Row row = sheet.createRow(r + 1);
                for (int c = 0; c < sampleData[r].length; c++) {
                    Cell cell = row.createCell(c);
                    cell.setCellValue((String) sampleData[r][c]);
                }
            }

            for (int i = 0; i < HEADERS.length; i++) {
                sheet.autoSizeColumn(i);
            }

            workbook.write(out);
            return out.toByteArray();

        } catch (Exception e) {
            log.error("Failed to generate master funds Excel template", e);
            throw new RuntimeException("Could not generate master funds Excel template", e);
        }
    }

    /**
     * Parses an uploaded Excel (.xlsx, .xls) spreadsheet into a list of MasterFundRowDto records.
     */
    public List<MasterFundRowDto> parseMasterFundsExcel(InputStream is) {
        List<MasterFundRowDto> list = new ArrayList<>();

        try (Workbook workbook = WorkbookFactory.create(is)) {
            Sheet sheet = workbook.getSheetAt(0);
            if (sheet == null) {
                throw new BadRequestException("The uploaded spreadsheet does not contain any sheets");
            }

            for (int i = 1; i <= sheet.getLastRowNum(); i++) {
                Row row = sheet.getRow(i);
                if (row == null || isRowEmpty(row)) {
                    continue;
                }

                String fundName = getCellValue(row, 0);
                String isin = getCellValue(row, 1);
                String categoryStr = getCellValue(row, 2);
                String subCategory = getCellValue(row, 3);
                String assetClass = getCellValue(row, 4);
                String instrumentType = getCellValue(row, 5);
                String activeStr = getCellValue(row, 6);

                if (fundName == null || fundName.isBlank()) {
                    log.warn("Row {} skipped: missing fund name", i + 1);
                    continue;
                }
                if (isin == null || isin.isBlank()) {
                    log.warn("Row {} skipped: missing ISIN for fund '{}'", i + 1, fundName);
                    continue;
                }
                if (categoryStr == null || categoryStr.isBlank()) {
                    log.warn("Row {} skipped: missing score category for fund '{}'", i + 1, fundName);
                    continue;
                }

                ScoreCategory category;
                try {
                    category = ScoreCategory.fromCode(categoryStr.trim());
                } catch (IllegalArgumentException e) {
                    log.warn("Row {} skipped: invalid score category '{}'", i + 1, categoryStr);
                    continue;
                }

                boolean isActive = activeStr == null || activeStr.isBlank() || "true".equalsIgnoreCase(activeStr.trim());
                String resolvedInstrument = (instrumentType != null && !instrumentType.isBlank()) ? instrumentType.trim() : "Mutual Fund";

                list.add(new MasterFundRowDto(
                        fundName.trim(),
                        isin.trim().toUpperCase(),
                        category,
                        subCategory != null ? subCategory.trim() : "",
                        assetClass != null ? assetClass.trim() : "Equity",
                        resolvedInstrument,
                        isActive
                ));
            }

        } catch (BadRequestException bre) {
            throw bre;
        } catch (Exception e) {
            log.error("Failed to parse master funds Excel file", e);
            throw new BadRequestException("Could not parse master funds file: " + e.getMessage());
        }

        return list;
    }

    private String getCellValue(Row row, int cellIndex) {
        Cell cell = row.getCell(cellIndex);
        if (cell == null || cell.getCellType() == CellType.BLANK) {
            return null;
        }
        return dataFormatter.formatCellValue(cell).trim();
    }

    private boolean isRowEmpty(Row row) {
        for (int c = row.getFirstCellNum(); c < row.getLastCellNum(); c++) {
            Cell cell = row.getCell(c);
            if (cell != null && cell.getCellType() != CellType.BLANK) {
                String val = dataFormatter.formatCellValue(cell).trim();
                if (!val.isEmpty()) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: backend-nodejs/src/modules/portfolioreview/services/eligibleFundExcelService.ts
import ExcelJS from 'exceljs';
import { AppError } from '../../../common/utils/AppError';
import { logger } from '../../../common/utils/logger';
import { s3Service } from '../../../common/services/s3Service';
import { EligibleFund } from '../models/EligibleFund';
import { ScoreCategoryCode } from '../../riskappetite/enums/riskEnums';
import { MasterFundUploadResponseDto } from '../dto/portfolioDto';

interface ParsedFundRow {
  fundName: string;
  isin: string;
  fundSubCategory: string;
  assetClass: string;
  instrumentType: string;
  scoreCategory: ScoreCategoryCode;
  isActive: boolean;
}

export class EligibleFundExcelService {
  /**
   * Generates a pre-styled .xlsx template for Master Funds upload with sample rows across all risk appetite categories.
   */
  async generateTemplate(): Promise<Buffer> {
    const workbook = new ExcelJS.Workbook();
    const worksheet = workbook.addWorksheet('Master Funds Template');

    worksheet.columns = [
      { header: 'Fund Name', key: 'fundName', width: 45 },
      { header: 'ISIN', key: 'isin', width: 18 },
      { header: 'Score Category', key: 'scoreCategory', width: 22 },
      { header: 'Sub Category', key: 'fundSubCategory', width: 28 },
      { header: 'Asset Class', key: 'assetClass', width: 16 },
      { header: 'Instrument Type', key: 'instrumentType', width: 18 },
      { header: 'Active (TRUE/FALSE)', key: 'isActive', width: 20 },
    ];

    // Header styling
    const headerRow = worksheet.getRow(1);
    headerRow.font = { bold: true, color: { argb: 'FFFFFFFF' } };
    headerRow.fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FF004D40' }, // Dark teal to match enterprise branding
    };
    headerRow.alignment = { horizontal: 'center', vertical: 'middle' };

    // Sample data rows across 5 Score Categories
    const sampleRows = [
      {
        fundName: 'ICICI Prudential Liquid Fund - Direct Growth',
        isin: 'INF109K01BE1',
        scoreCategory: 'VERY_CONSERVATIVE',
        fundSubCategory: 'Liquid',
        assetClass: 'Debt',
        instrumentType: 'Mutual Fund',
        isActive: 'TRUE',
      },
      {
        fundName: 'SBI Balanced Advantage Fund - Direct Growth',
        isin: 'INF200K01BE4',
        scoreCategory: 'CONSERVATIVE',
        fundSubCategory: 'Dynamic Asset Allocation',
        assetClass: 'Hybrid',
        instrumentType: 'Mutual Fund',
        isActive: 'TRUE',
      },
      {
        fundName: 'HDFC Top 100 Fund - Direct Growth',
        isin: 'INF179K01BE2',
        scoreCategory: 'MODERATE',
        fundSubCategory: 'Large Cap',
        assetClass: 'Equity',
        instrumentType: 'Mutual Fund',
        isActive: 'TRUE',
      },
      {
        fundName: 'Parag Parikh Flexi Cap Fund - Direct Growth',
        isin: 'INF879O01019',
        scoreCategory: 'AGGRESSIVE',
        fundSubCategory: 'Flexi Cap',
        assetClass: 'Equity',
        instrumentType: 'Mutual Fund',
        isActive: 'TRUE',
      },
      {
        fundName: 'Nippon India Small Cap Fund - Direct Growth',
        isin: 'INF204K01BE3',
        scoreCategory: 'VERY_AGGRESSIVE',
        fundSubCategory: 'Small Cap',
        assetClass: 'Equity',
        instrumentType: 'Mutual Fund',
        isActive: 'TRUE',
      },
    ];

    for (const sample of sampleRows) {
      worksheet.addRow(sample);
    }

    const uint8Array = await workbook.xlsx.writeBuffer();
    return Buffer.from(uint8Array);
  }

  /**
   * Normalizes category string to valid ScoreCategoryCode enum.
   */
  private normalizeCategory(rawCategory: string): ScoreCategoryCode | null {
    const cleaned = rawCategory.trim().toLowerCase().replace(/[\s-]+/g, '_');
    const validCategories = Object.values(ScoreCategoryCode);
    if (validCategories.includes(cleaned as ScoreCategoryCode)) {
      return cleaned as ScoreCategoryCode;
    }
    return null;
  }

  /**
   * Parses uploaded Excel file, archives file in AWS S3 / LocalStack, and upserts funds into MongoDB by ISIN.
   */
  async processUpload(buffer: Buffer, originalFilename: string): Promise<MasterFundUploadResponseDto> {
    const safeFilename = originalFilename.replace(/[^a-zA-Z0-9._-]/g, '_');
    const s3Key = `master-funds/${Date.now()}_${safeFilename}`;
    let presignedUrl: string | null = null;

    // 1. Upload raw snapshot to AWS S3 / LocalStack
    try {
      await s3Service.uploadFile({
        key: s3Key,
        buffer,
        contentType: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      });
      presignedUrl = await s3Service.getPresignedDownloadUrl(s3Key);
    } catch (s3Error: unknown) {
      const msg = s3Error instanceof Error ? s3Error.message : String(s3Error);
      logger.warn(
        { error: msg, s3Key },
        'S3 upload failed or LocalStack unavailable. Continuing with database ingestion.'
      );
    }

    // 2. Parse Excel workbook
    const workbook = new ExcelJS.Workbook();
    try {
      await workbook.xlsx.load(buffer as unknown as Parameters<typeof workbook.xlsx.load>[0]);
    } catch (parseError: unknown) {
      const parseMsg = parseError instanceof Error ? parseError.message : String(parseError);
      logger.warn({ error: parseMsg, originalFilename }, 'Failed to parse Excel file workbook');
      throw new AppError(`Could not parse master funds Excel file: ${parseMsg}`, 400);
    }

    const worksheet = workbook.worksheets[0];
    if (!worksheet || worksheet.rowCount < 2) {
      logger.warn({ originalFilename }, 'Master funds upload failed: Excel sheet is empty');
      throw new AppError('The uploaded spreadsheet does not contain any data rows', 400);
    }

    // Dynamically detect column mapping from header row
    const headerRow = worksheet.getRow(1);
    let colFundName = 1;
    let colIsin = 2;
    let colScoreCategory = 3;
    let colSubCategory = 4;
    let colAssetClass = 5;
    let colInstrumentType = 6;
    let colActive = 7;

    headerRow.eachCell((cell, colNumber) => {
      const headerText = String(cell.text || '').toLowerCase().trim();
      if (headerText.includes('fund name') || headerText.includes('scheme name')) {
        colFundName = colNumber;
      } else if (headerText.includes('isin')) {
        colIsin = colNumber;
      } else if (headerText.includes('score category') || headerText === 'category' || headerText.includes('risk category')) {
        colScoreCategory = colNumber;
      } else if (headerText.includes('sub category') || headerText.includes('subcategory')) {
        colSubCategory = colNumber;
      } else if (headerText.includes('asset class')) {
        colAssetClass = colNumber;
      } else if (headerText.includes('instrument')) {
        colInstrumentType = colNumber;
      } else if (headerText.includes('active')) {
        colActive = colNumber;
      }
    });

    const fundsToUpsert: ParsedFundRow[] = [];

    worksheet.eachRow((row, rowNumber) => {
      // Skip header row
      if (rowNumber === 1) return;

      const fundName = String(row.getCell(colFundName).text || '').trim();
      const isin = String(row.getCell(colIsin).text || '').trim().toUpperCase();
      const rawCategory = String(row.getCell(colScoreCategory).text || '').trim();
      const fundSubCategory = String(row.getCell(colSubCategory).text || '').trim();
      const assetClass = String(row.getCell(colAssetClass).text || 'Equity').trim() || 'Equity';
      const instrumentType = String(row.getCell(colInstrumentType).text || 'Mutual Fund').trim() || 'Mutual Fund';
      const activeText = String(row.getCell(colActive).text || 'TRUE').trim().toUpperCase();

      // Skip completely blank rows
      if (!isin && !fundName) return;

      if (!fundName) {
        logger.warn({ rowNumber }, 'Master fund row skipped: Missing fund name');
        return;
      }

      if (!isin) {
        logger.warn({ rowNumber, fundName }, 'Master fund row skipped: Missing ISIN');
        return;
      }

      if (!rawCategory) {
        logger.warn({ rowNumber, fundName, isin }, 'Master fund row skipped: Missing score category');
        return;
      }

      const scoreCategory = this.normalizeCategory(rawCategory);
      if (!scoreCategory) {
        logger.warn(
          { rowNumber, fundName, rawCategory, validCategories: Object.values(ScoreCategoryCode) },
          'Master fund row skipped: Invalid score category'
        );
        return;
      }

      fundsToUpsert.push({
        fundName,
        isin,
        fundSubCategory: fundSubCategory || 'Diversified',
        assetClass,
        instrumentType,
        scoreCategory,
        isActive: activeText !== 'FALSE',
      });
    });

    if (fundsToUpsert.length === 0) {
      logger.warn({ originalFilename }, 'Master funds upload failed: No valid fund rows found');
      throw new AppError('No valid fund records found in the uploaded file', 400);
    }

    // 3. Upsert into MongoDB by ISIN
    let insertedRecords = 0;
    let updatedRecords = 0;

    for (const fund of fundsToUpsert) {
      const existing = await EligibleFund.findOne({ isin: fund.isin });
      if (existing) {
        existing.fundName = fund.fundName;
        existing.fundSubCategory = fund.fundSubCategory;
        existing.assetClass = fund.assetClass;
        existing.instrumentType = fund.instrumentType;
        existing.scoreCategory = fund.scoreCategory;
        existing.isActive = fund.isActive;
        await existing.save();
        updatedRecords++;
      } else {
        await EligibleFund.create(fund);
        insertedRecords++;
      }
    }

    logger.info(
      { totalParsed: fundsToUpsert.length, insertedRecords, updatedRecords, s3Key },
      'Master funds successfully processed and upserted into database'
    );

    return {
      status: 'SUCCESS',
      message: `Master funds processed successfully (${insertedRecords} inserted, ${updatedRecords} updated)`,
      filename: originalFilename,
      s3Key,
      fileUrl: presignedUrl,
      totalRecords: fundsToUpsert.length,
      insertedRecords,
      updatedRecords,
    };
  }
}

export const eligibleFundExcelService = new EligibleFundExcelService();
